core/modules/link/link.module

```diff
@@ -849,7 +866,13 @@ function _link_default_attributes() {
  * The $fields array is in $complete_form['#field_info'][$element['#field_name']].
  */
 function link_field_process($element, $form_state, $complete_form) {
-  $instance = field_widget_instance($element, $form_state);
+  $field_parents = (!empty($element['#field_parents'])) ? $element['#field_parents'] : array();
+  $field_name = (!empty($element['#field_name'])) ? $element['#field_name'] : '';
+  $language = (!empty($element['#language'])) ? $element['#language'] : LANGUAGE_NONE;
+  $field_state = field_form_get_state($field_parents, $field_name, $language, $form_state);
+  $instance = !empty($field_state['instance']) ? $field_state['instance'] : link_field_custom_create_instance($element);
```

Comment:
The idea of having to create a dummy field instance for this to work seems strange to me. Elements are low level form API elements whose functionality should not depend on Field API. It's probably more work, but I would be inclined to refactor the whole thing to clear separate form element functionality and Field API widget one (which is the only one who should be concerned with field instances).

I realize this may block this issue for some time, but I'm thinking that refactoring things will result on more streamlined code that follows the pattern of other complex elements / fields, such as date elements and date fields.

Reply:
@argiepiano I'm not against your idea. Initially my goal was just to prevent a series of warnings that showed up when trying to use `link_field` in a custom form, and it's kind of snowballing into a bigger project. I'm out of time on this one for now but I'm sure your idea is a better approach from a top-level perspective.

Reply:
That makes sense. Let me see if I can do some testing as is, and perhaps suggest a different approach over the weekend.

Reply:
Closing as outdated, but feedback on current implementation would be great if you get a chance.